### PR TITLE
Streams with flexible requestN semantic

### LINKED ISSUE #109: Limit rate operator for streams
Need to implement some limit rate operator for streams and channels, to improve flexibility of streams API.
F.e to support similar operator from reactor like [Flux.limitRate](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html#limitRate-int-int-) or reactor-like prefetch logic: request new elements, when 75% of elements collected

### PR DESCRIPTION
**Pull request by whyoleg**
* Reworked streams API using coroutine context element RequestStrategy for better control over requestN frames
* fixes #109

### CODE REVIEW

**Comment by whyoleg** [comment]
@yschimke
What do you think about it?
I've found, that it's really not so easy to integrate better backpressure control using just Flow and current RSocket interface. So I created new `ReactiveFlow` inherited from `Flow` (naming is TBD). And because only incoming Flow should be controlled by requestN (so on requester side it's requestStream and requestChannel output and on responder side requestChannel input), and for other cases it's not needed, and will just add complexite, I've decided to split RSocket interface for requester and responder.
So, do you think, that this approach is OK, and what do you think about `RequestStrategy` API?

and yes, it's still WIP, need more tests and fixes may be, but overall works

**Comment by yschimke** [comment]
@whyoleg I think I'm the wrong reviewer for this.  You should try to get some of the Pivotal folk with more experience across rsocket-java and spring reactor. There have been a lot of discussions in the past on reactors infinite and cancel defaults and limits, and I have simple requirements to be honest now.

In FB days it was more relevant to me since there was a polling mechanism in the background and a real cost to every additional emmitted result, but infinite and cancel is mostly fine for what I'm doing now e.g. rsocket-demo with a stream of tweets (line separated http response) or with my own backends being purely event based anyway.

File: benchmarks/src/kotlinMain/kotlin/io/rsocket/kotlin/benchmarks/RSocketKotlinBenchmark.kt [hunk 1]
```diff
@@ -27,31 +27,33 @@ import kotlin.random.*
...
 class RSocketKotlinBenchmark : RSocketBenchmark<Payload>() {
 
-    lateinit var client: RSocket
+    lateinit var client: RSocketRequester
     lateinit var server: Job
 
     lateinit var payload: Payload
     lateinit var payloadsFlow: Flow<Payload>
 
+    fun payloadCopy(): Payload = payload.copy()
```
Comment (yschimke) [inline benchmarks/src/kotlinMain/kotlin/io/rsocket/kotlin/benchmarks/RSocketKotlinBenchmark.kt:36]: This sort of pattern, reuse of a payload probably deserves some nicer more natural mechanism.  This is likely a source of bugs under error conditions which will be missed on the happy path.  How do we improve this? e.g. payload.permanent()?

**Comment by yschimke** [comment]
Or maybe @elizarov and work out what the extension methods to ReactiveFlow should be.  I think ideally none new inside rsocket-kotlin.  We would just react to the usage and contract of ReactiveFlow.  And use concepts the end users are already familiar with externally.

IIRC That was essentially the decision with rsocket-java and spring reactor. Not to build our own terminology, but to keep using the operators from the underlying framework.

File: benchmarks/src/kotlinMain/kotlin/io/rsocket/kotlin/benchmarks/RSocketKotlinBenchmark.kt (hunk 1 above)
Comment (whyoleg) [reply benchmarks/src/kotlinMain/kotlin/io/rsocket/kotlin/benchmarks/RSocketKotlinBenchmark.kt:36]: and how `payload.permanent()` will work? have just possibility to reuse data? and release after all?
@altavir has much experience working with kotlin with IO, and now researching adding rsocket-kotlin to his own projects. Maybe hi have some ideas on better payload handling?

**Comment by whyoleg** [comment]
The idea behind `ReactiveFlow` is to just store some `RequestStrategy` which will allow for better control of requestN frames and in current implementation, it has only one realization `PrefetchStrategy(requestSize, requestOn)` - which will request `requestSize` elements, when `requestOn` elements are left for collecting in `Flow`. So, the simplest one.
Im not sure, that something else should be added for now. From user perspective, it's even may be not needed, and default strategy (requestSize=64, requestOn=16) will be enough.

We need to add such logic, with one new type (`ReactiveFlow`), as `Flow` only works with collecting one by one, and even `take(5)` will work as collect by one until 5 elements collected.

**Comment by yschimke** [comment]
I'm not a huge fan of defining our own ReactiveFlow

**Comment by elizarov** [comment]
If all you need is to store a request strategy, then I can suggest looking into using a context for this kind of storage. This way you can have a sensible default that will add zero overhead (and zero new concepts) and rely on the existing context-passing support in both Reactor and Kotlin Flow to change this request strategy if needed.

**Comment by whyoleg** [comment]
@elizarov cool idea, I will try to do that! thx!

**Comment by OlegDokuka** [comment]
as discussed, lets split these changes into two PRs:
1) For RequestN via `flowOn` 
2) RSocket interface API changes

File: benchmarks/src/kotlinMain/kotlin/io/rsocket/kotlin/benchmarks/RSocketKotlinBenchmark.kt (hunk 1 above)
Comment (OlegDokuka) [reply benchmarks/src/kotlinMain/kotlin/io/rsocket/kotlin/benchmarks/RSocketKotlinBenchmark.kt:38]: Just from my side. We do allocate a ByteBuf once, and then retain it every time we use it in benchmarks. Sounds like a copy but not really, thus not sure Kotlin IO can do the same

File: rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        require(requestOn in 0 until requestSize) { "requestSize and requestOn should be in relation: requestSize > requestOn >= 0" }
+    }
+
+    override fun provide(): RequestStrategy.Element = Element(requestSize, requestOn)
+
+    private class Element(
+        private val requestSize: Int,
+        private val requestOn: Int,
+    ) : RequestStrategy.Element {
+        private val requested = atomic(requestSize)
```
Comment (OlegDokuka) [inline rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt:64]: Just wondering why do we need atomic here?

File: rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        override suspend fun firstRequest(): Int = requestSize
+
+        override suspend fun nextRequest(): Int {
+            if (requested.decrementAndGet() != requestOn) return 0
+
+            requested += requestSize
```
Comment (OlegDokuka) [inline rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt:70]: but here, this atomic is used in non-atomic way, Am I missing something?

**Review by OlegDokuka — CHANGES_REQUESTED** [review]
LGTM with some minor comments

File: rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt (hunk 2 above)
Comment (whyoleg) [reply rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt:64]: it needed to allow mutation on K/N if we freeze. need to check, may be it's really not needed, here.
On K/JVM and K/JS it's not needed, as functions inside `RequestStrategy.Element` will never be called concurrently

Comment (OlegDokuka) [reply rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt:64]: yes, please double-check. In general, it looks like all the requests will be done within the same context, hence the context should do all the stuff to keep seriality, hence no atomic is needed

Comment (whyoleg) [reply rsocket-core/src/commonMain/kotlin/io/rsocket/kotlin/RequestStrategy.kt]: atomic removed, not needed, thx for pointing!

**Review by OlegDokuka — APPROVED** [review]
LGTM 🔥

**Pull request merged**